src/v/strings/static_str.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include <fmt/format.h>
+
+#include <string_view>
+#include <type_traits>
+
+#pragma once
+
+/**
+ * @brief A wrapper around a compile-time constant string.
+ *
+ * A static_str wraps an compile-time constant string, and is intended to be
+ * used at runtime as a guarantee that a string has static lifetime. Such string
+ * have trivial and cheap copy operations, since we just copy pointers (or
+ * views) to the underlying string.
+ */
+class static_str {
+public:
+    /**
+     * @brief Create a static_str based on a compile-time string literal.
+     *
+     * The use of consteval prevents passing anything other than compile-time
+     * values, which necessarily have static lifetime.
+     */
+    consteval static_str(const char* str) // NOLINT(hicpp-explicit-conversions)
+      : _data{str} {}
+
+    /**
+     * @brief Create a static_str based on a compile-time string literal.
+     *
+     * The use of consteval prevents passing anything other than compile-time
+     * values, which necessarily have static lifetime.
+     */
+    explicit consteval static_str(std::string_view str)
+      : _data{str} {}
+
+    constexpr bool operator==(const static_str& rhs) const {
+        return same_pointer(rhs) || _data == rhs._data;
+    }
+
+    constexpr auto operator<=>(const static_str& rhs) const {
+        if (same_pointer(rhs)) {
+            return std::strong_ordering::equal;
+        }
+        return _data <=> rhs._data;
+    }
+
+    /**
+     * @brief Return the static_str as a string_view.
+     */
+    constexpr operator std::string_view() const { return _data; } // NOLINT
+
+private:
+    // Return true if this string and rhs live at the same address. This always
+    // returns false if evaluated at constexpr time because constexpr must not
+    // do this comparison. This idea is to use it on a fast-path for == and <=>
+    // because pointer equality is a sufficient but not necessary condition for
+    // string equality.
+    constexpr bool same_pointer(const static_str& rhs) const {
```

Comment:
Do we need to compare end of the the string too or are we sure that we won't get overlapping substrings?

Reply:
Good catch. This _is_ possible in the most recent version of this PR because we switched to using std::string_view as the storage and also just now added a constructor taking string_view, which opens the possibility of storing a view which does not extend to null terminator so yes this prefix bug is possible.

In push xxx I've just remove this optimization entirely and defaulted <=> operator.

Reply:
Any type of minimal fuzzer would have caught this bug, sigh.